include/kernels/gauge_wilson_flow.cuh

```diff
@@ -56,48 +56,65 @@ namespace quda
     }
   };
 
-  template <typename Arg>
-  __host__ __device__ inline auto computeStaple(const Arg &arg, const int *x, int parity, int dir)
+  template <typename Arg> struct computeStapleOpsWF {
+    using real = typename Arg::real;
+    using Link = Matrix<complex<real>, Arg::nColor>;
+    using WilsonOps = computeStapleOps;
+    using StapOp = ThreadLocalCache<Link, 0, computeStapleRectangleOps>; // offset by computeStapleRectangleOps
+    using RectOp = ThreadLocalCache<Link, 0, StapOp>;                    // offset by StapOp
+    using SymanzikOps = combineOps<computeStapleRectangleOps, KernelOps<StapOp, RectOp>>;
+    using Ops = std::conditional_t<Arg::wflow_type == QUDA_GAUGE_SMEAR_SYMANZIK_FLOW, SymanzikOps, WilsonOps>;
+  };
+
+  template <typename Ftor>
+  __host__ __device__ inline auto computeStaple(const Ftor &ftor, const int *x, int parity, int dir)
   {
+    using Arg = typename Ftor::Arg;
+    const Arg &arg = ftor.arg;
     using real = typename Arg::real;
     using Link = Matrix<complex<real>, Arg::nColor>;
     Link Z;
     // Compute staples and Z factor
-    switch (arg.wflow_type) {
-    case QUDA_GAUGE_SMEAR_WILSON_FLOW :
+    static_assert(Arg::wflow_type == QUDA_GAUGE_SMEAR_WILSON_FLOW || Arg::wflow_type == QUDA_GAUGE_SMEAR_SYMANZIK_FLOW);
+    if constexpr (Arg::wflow_type == QUDA_GAUGE_SMEAR_WILSON_FLOW) {
       // This function gets stap = S_{mu,nu} i.e., the staple of length 3,
-      computeStaple(arg, x, arg.E, parity, dir, Z, Arg::wflow_dim);
-      break;
-    case QUDA_GAUGE_SMEAR_SYMANZIK_FLOW :
+      computeStaple(ftor, x, arg.E, parity, dir, Z, Arg::wflow_dim);
+    }
```

Comment:
Pedantic, but I think it would be better practice to have an `else if constexpr` here. I know it's essentially impossible for the conditions in each of the `if` statements to both be true, this is just a thought on a convention.

Reply:
6a727a700e87b08f30a3874877cd11e927b95f90